RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1TkMu.cc

```diff
@@ -0,0 +1,413 @@
+    theOfflineSeedLabel = iConfig.getUntrackedParameter<InputTag>("OfflineSeedLabel");
+    offlineSeedToken_ = consumes<edm::View<TrajectorySeed>>(theOfflineSeedLabel);
+
+    // check that number of eta bins -1 matches number of dR cones
+    if (matchingDR.size() != etaBins.size() - 1) {
+      throw cms::Exception("Configuration") << "Size of MatchDR "
+                                            << "does not match number of eta bins." << endl;
+    }
+  }
+
+  // service parameters
+  ParameterSet serviceParameters = iConfig.getParameter<ParameterSet>("ServiceParameters");
+
+  // the services
+  theService = new MuonServiceProxy(serviceParameters, consumesCollector());
+
+  // the estimator
+  theEstimator = new Chi2MeasurementEstimator(10000.);
+
+  produces<L2MuonTrajectorySeedCollection>();
+}
+
+// destructor
+L2MuonSeedGeneratorFromL1TkMu::~L2MuonSeedGeneratorFromL1TkMu() {
+  if (theService)
+    delete theService;
+  if (theEstimator)
+    delete theEstimator;
+}
+
+void L2MuonSeedGeneratorFromL1TkMu::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("InputObjects", edm::InputTag("hltGmtStage2Digis"));
+  desc.add<string>("Propagator", "");
+  desc.add<double>("L1MinPt", -1.);
+  desc.add<double>("L1MaxEta", 5.0);
+  desc.add<double>("SetMinPtBarrelTo", 3.5);
+  desc.add<double>("SetMinPtEndcapTo", 1.0);
+  desc.add<double>("MinPL1Tk", 3.5);
+  desc.add<double>("MinPtL1TkBarrel", 3.5);
+  desc.addUntracked<bool>("UseOfflineSeed", false);
+  desc.add<bool>("UseUnassociatedL1", true);
+  desc.add<std::vector<double>>("MatchDR", {0.3});
+  desc.add<std::vector<double>>("EtaMatchingBins", {0., 2.5});
+  desc.addUntracked<edm::InputTag>("OfflineSeedLabel", edm::InputTag(""));
+
+  edm::ParameterSetDescription psd0;
+  psd0.addUntracked<std::vector<std::string>>("Propagators", {"SteppingHelixPropagatorAny"});
+  psd0.add<bool>("RPCLayers", true);
+  psd0.addUntracked<bool>("UseMuonNavigation", true);
+  desc.add<edm::ParameterSetDescription>("ServiceParameters", psd0);
+  descriptions.add("L2MuonSeedGeneratorFromL1TkMu", desc);
+}
+
+void L2MuonSeedGeneratorFromL1TkMu::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  const std::string metname = "Muon|RecoMuon|L2MuonSeedGeneratorFromL1TkMu";
+  MuonPatternRecoDumper debug;
+
+  auto output = std::make_unique<L2MuonTrajectorySeedCollection>();
+
+  edm::Handle<l1t::TkMuonCollection> muColl;
+  iEvent.getByToken(muCollToken_, muColl);
+  LogDebug(metname) << "Number of muons " << muColl->size() << endl;
+
+  edm::Handle<edm::View<TrajectorySeed>> offlineSeedHandle;
+  vector<int> offlineSeedMap;
+  if (useOfflineSeed) {
+    iEvent.getByToken(offlineSeedToken_, offlineSeedHandle);
+    offlineSeedMap = vector<int>(offlineSeedHandle->size(), 0);
+  }
+
+  for (auto ittkmu = muColl->begin(); ittkmu != muColl->end(); ittkmu++) {
+    // L1 tracker track
+    auto it = ittkmu->trkPtr();
```

Comment:
this could be simplified with a range-based loop:
```suggestion
  for (auto const& tkmu : muColl)) {
    // L1 tracker track
    auto it = tkmu.trkPtr();
```